Protocol doesn't inherit @MainActor requirement from base class

**Description**

When a protocol inherits from a `@MainActor` class, I would expect it to inherit the `@MainActor` restriction, but it does not.

**Steps to reproduce**
```swift
@MainActor
class MAC {
    var s: String = "hi"
}

protocol P: MAC {
    var t: String { get }
}

class X: MAC, P {
    var t: String {
        s
    }
}
```

`swift -strict-concurrency=complete mainactor-protocol-from-class.swift`

**Expected behavior**

Code compiles.

**Actual behavior**

```
mainactor-protocol-from-class.swift:11:9: warning: main actor-isolated property 't' cannot be used to satisfy nonisolated protocol requirement
    var t: String {
        ^
mainactor-protocol-from-class.swift:7:9: note: 't' declared here
    var t: String { get }
        ^
```

(It makes more sense that it does not compile if you consider that `protocol P: MAC` is syntactic sugar for `protocol P where Self: MAC`, but it looks like an inheritance clause, so I expect global actor inheritance to work as it does in other contexts)

You can add `@MainActor` explicitly to `P` to make the code compile.

**Environment**
- swift-driver version: 1.62.15 Apple Swift version 5.7.2 (swiftlang-5.7.2.135.5 clang-1400.0.29.51)
- Target: arm64-apple-macosx13.0
- Xcode 14.2
- Build version 14C18

--- a/lib/Sema/TypeCheckConcurrency.cpp
+++ b/lib/Sema/TypeCheckConcurrency.cpp
@@ -4603,13 +4603,15 @@ getIsolationFromConformances(NominalTypeDecl *nominal) {
 static std::optional<ActorIsolation>
 getIsolationFromInheritedProtocols(ProtocolDecl *protocol) {
   std::optional<ActorIsolation> foundIsolation;
-  for (auto inherited : protocol->getInheritedProtocols()) {
-    switch (auto protoIsolation = getActorIsolation(inherited)) {
+  bool conflict = false;
+
+  auto inferIsolation = [&](ValueDecl *decl) {
+    switch (auto protoIsolation = getActorIsolation(decl)) {
     case ActorIsolation::ActorInstance:
     case ActorIsolation::Unspecified:
     case ActorIsolation::Nonisolated:
     case ActorIsolation::NonisolatedUnsafe:
-      break;
+      return;
 
     case ActorIsolation::Erased:
       llvm_unreachable("protocol cannot have erased isolation");
@@ -4617,15 +4619,26 @@ getIsolationFromInheritedProtocols(ProtocolDecl *protocol) {
     case ActorIsolation::GlobalActor:
       if (!foundIsolation) {
         foundIsolation = protoIsolation;
-        continue;
+        return;
       }
 
       if (*foundIsolation != protoIsolation)
-        return std::nullopt;
+        conflict = true;
 
-      break;
+      return;
     }
+  };
+
+  for (auto inherited : protocol->getInheritedProtocols()) {
+    inferIsolation(inherited);
   }
+
+  if (auto *superclass = protocol->getSuperclassDecl()) {
+    inferIsolation(superclass);
+  }
+
+  if (conflict)
+    return std::nullopt;
 
   return foundIsolation;
 }

--- a/test/Concurrency/global_actor_inference_swift6.swift
+++ b/test/Concurrency/global_actor_inference_swift6.swift
@@ -2,7 +2,7 @@
 
 // RUN: %target-swift-frontend -swift-version 6 -emit-module -emit-module-path %t/other_global_actor_inference.swiftmodule -module-name other_global_actor_inference -strict-concurrency=complete %S/Inputs/other_global_actor_inference.swift
 
-// RUN: %target-swift-frontend -swift-version 6 -I %t -disable-availability-checking %s -emit-sil -o /dev/null -verify -enable-upcoming-feature GlobalActorIsolatedTypesUsability
+// RUN: %target-swift-frontend -swift-version 6 -I %t -disable-availability-checking %s -emit-sil -o /dev/null -verify
 
 // REQUIRES: concurrency
 
@@ -179,3 +179,37 @@ extension S3 {
   // expected-error@-1{{global actor 'SomeGlobalActor'-isolated instance method 'f()' cannot be used to satisfy main actor-isolated protocol requirement}}
   //expected-note@-2{{add 'nonisolated' to 'f()' to make this instance method not isolated to the actor}}
 }
+
+@MainActor
+func onMain() {}
+
+@MainActor
+class MainActorSuperclass {}
+
+protocol InferMainFromSuperclass: MainActorSuperclass {
+  func f()
+}
+
+class C1: MainActorSuperclass, InferMainFromSuperclass {
+  func f() {
+    onMain() // okay
+  }
+}
+
+protocol InferenceConflictWithSuperclass: MainActorSuperclass, InferSomeGlobalActor {
+  func g()
+  // expected-note@-1 {{mark the protocol requirement 'g()' 'async' to allow actor-isolated conformances}}
+}
+
+
+class C2: MainActorSuperclass, InferenceConflictWithSuperclass {
+//expected-note@-1 {{add '@preconcurrency' to the 'InferenceConflictWithSuperclass' conformance to defer isolation checking to run time}}
+
+  func f() {}
+
+  func g() {}
+  // expected-error@-1 {{main actor-isolated instance method 'g()' cannot be used to satisfy nonisolated protocol requirement}}
+  // expected-note@-2 {{add 'nonisolated' to 'g()' to make this instance method not isolated to the actor}}
+}
+
+

